packages/firestore/test/integration/api_internal/bundle.test.ts

```diff
@@ -0,0 +1,240 @@
+
+      // Read from cache. These documents do not exist in backend, so they can
+      // only be read from cache.
+      const snap = await db.collection('coll-1').get({ source: 'cache' });
+      verifySnapEqualTestDocs(snap);
+    });
+  });
+
+  it('load for a second time skips', () => {
+    return withTestDb(persistence, async db => {
+      const builder = bundleWithTestDocs(db);
+
+      await db.loadBundle(
+        builder.build('test-bundle', { seconds: 1001, nanos: 9999 })
+      );
+
+      let completeCalled = false;
+      const progressEvents: firestore.LoadBundleTaskProgress[] = [];
+      await db
+        .loadBundle(
+          encoder.encode(
+            builder.build('test-bundle', { seconds: 1001, nanos: 9999 })
+          )
+        )
+        .onProgress(
+          progress => {
+            progressEvents.push(progress);
+          },
+          error => {},
+          () => {
+            completeCalled = true;
+          }
+        );
+
+      expect(completeCalled).to.be.true;
+      // No loading actually happened in the second `loadBundle` call only the
+      // success progress is recorded.
+      expect(progressEvents.length).to.equal(1);
+      verifySuccessProgress(progressEvents[0]);
+
+      // Read from cache. These documents do not exist in backend, so they can
+      // only be read from cache.
+      const snap = await db.collection('coll-1').get({ source: 'cache' });
+      verifySnapEqualTestDocs(snap);
+    });
+  });
+
+  it('load with documents already pulled from backend', () => {
+    return withTestDb(persistence, async db => {
+      await db.doc('coll-1/a').set({ k: 'a', bar: 0 });
+      await db.doc('coll-1/b').set({ k: 'b', bar: 0 });
+
+      const accumulator = new EventsAccumulator<firestore.QuerySnapshot>();
+      db.collection('coll-1').onSnapshot(accumulator.storeEvent);
+      await accumulator.awaitEvent();
+
+      const builder = bundleWithTestDocs(db);
+      const progress = await db.loadBundle(
+        // Testing passing in non-string bundles.
+        encoder.encode(
+          builder.build('test-bundle', { seconds: 1001, nanos: 9999 })
+        )
+      );
+
+      verifySuccessProgress(progress);
+      // The test bundle is holding ancient documents, so no events are
+      // generated as a result. The case where a bundle has newer doc than
+      // cache can only be tested in spec tests.
+      await accumulator.assertNoAdditionalEvents();
+
+      const snap = await db.collection('coll-1').get();
+      expect(toDataArray(snap)).to.deep.equal([
+        { k: 'a', bar: 0 },
+        { k: 'b', bar: 0 }
+      ]);
+    });
+  });
+
+  it('load with documents from other projects fails', () => {
+    return withTestDb(persistence, async db => {
+      let builder = bundleWithTestDocs(db);
+      return withAlternateTestDb(persistence, async otherDb => {
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        expect(
```

Comment:
This needs to be awaited, otherwise the assert can fire after the test completed.